pypesto/objective/amici_calculator.py

```diff
@@ -214,3 +221,54 @@ def calculate_function_values(rdatas,
         RDATAS: rdatas
     }
     return filter_return_dict(ret)
+
+
+def get_output(edatas: List['amici.ExpData'],
+               amici_model: AmiciModel,
+               amici_solver: AmiciSolver,
+               chunk_size: int,
+               mode: str,
+               n_threads: int):
+    rdatas = []
+    # spliting the ExpData in chunks and simulating chunk wise
+    # Do we have a maximum number of simulations allowed?
+    n_edatas = len(edatas)
+    if chunk_size is None:
+        # simulate all conditions at once
+        n_simulations = 1
+    else:
+        # simulate only a subset of conditions
+        n_simulations = int(np.ceil(len(edatas) / chunk_size))
+    # get keys based on mode
+    if mode == MODE_FUN:
+        requested_keys = ['llh', 'sllh', 'status']
+    elif mode == MODE_RES:
+        requested_keys = ['res', 'sres', 'status']
+    else:
+        requested_keys = None
```

Comment:
I think we need a small discussion perhaps here on what to keep and also add an option to keep the rdatas object as it was before?

Reply:
I think there were some reasons for always returning all values (llh, sllh, res, sres, ...) until some stage, for debugging or history creation. this would need to be evaluated carefully, whether reducing the output induces changes there.

Reply:
but it is also the crux as this is the thing that takes up the memory isn't it?

Reply:
I guess. What was the exact problem that you observed? Running into memory problems when there are many conditions?

Reply:
Yes. E.G. had 13000 conditions and got a SIGKILL. Activity monitor said that it required 200GB memory. Think @LeonardSchmiester had a similar problem.

Reply:
Agreed that one should then operate more efficiently, in particular the matrices sres, hess will be costly. @FFroehlich any opinions of what you would like to have always in the history / what would be good defaults?

Reply:
So in principle, I can revert everything back and look into the `amici_solver.setReturnDataReportingMode(...)`?
🔴 for no, 🟢 for yes.

Reply:
so I looked into the code and also in amici, but I do not quite understand how exactly it works.
I set the `RDataReporting=0` and `RDataReporting=2` with `amici_solver.setReturnDataReportingMode(...)`  but the dimensions of the rdatas I got were the same.

Reply:
> so I looked into the code and also in amici, but I do not quite understand how exactly it works.
> I set the `RDataReporting=0` and `RDataReporting=2` with `amici_solver.setReturnDataReportingMode(...)` but the dimensions of the rdatas I got were the same.

I think the fields will still get initialized, but some of them should contain empty arrays or None: https://github.com/AMICI-dev/AMICI/blob/master/src/rdata.cpp

Reply:
> 
> I think the fields will still get initialized, but some of them should contain empty arrays or None: https://github.com/AMICI-dev/AMICI/blob/master/src/rdata.cpp

That is also where I looked and what I thought, but as the dimensions of each fields were the same that would not resolve the issue would it? 

Here is a terminal output: `rdata_complete` had the solver set the `RDataReporting` to 0, while `data_in_chunks` had it set to 2.
```
(Pdb) rdata_complete['rdatas'][0]._field_dimensions
{'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
(Pdb) rdata_in_chunks['rdatas'][0]._field_dimensions
{'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
(Pdb)
```

Reply:
> > I think the fields will still get initialized, but some of them should contain empty arrays or None: https://github.com/AMICI-dev/AMICI/blob/master/src/rdata.cpp
> 
> That is also where I looked and what I thought, but as the dimensions of each fields were the same that would not resolve the issue would it?
> 
> Here is a terminal output: `rdata_complete` had the solver set the `RDataReporting` to 0, while `data_in_chunks` had it set to 2.
> 
> ```
> (Pdb) rdata_complete['rdatas'][0]._field_dimensions
> {'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
> (Pdb) rdata_in_chunks['rdatas'][0]._field_dimensions
> {'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
> (Pdb)
> ```

okay, thats not what's supposed to happend. Could be a bug in amici. Will have a look.

Reply:
> > > I think the fields will still get initialized, but some of them should contain empty arrays or None: https://github.com/AMICI-dev/AMICI/blob/master/src/rdata.cpp
> > 
> > 
> > That is also where I looked and what I thought, but as the dimensions of each fields were the same that would not resolve the issue would it?
> > Here is a terminal output: `rdata_complete` had the solver set the `RDataReporting` to 0, while `data_in_chunks` had it set to 2.
> > ```
> > (Pdb) rdata_complete['rdatas'][0]._field_dimensions
> > {'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
> > (Pdb) rdata_in_chunks['rdatas'][0]._field_dimensions
> > {'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
> > (Pdb)
> > ```
> 
> okay, thats not what's supposed to happen. Could be a bug in amici. Will have a look.

Reply:
> > I think the fields will still get initialized, but some of them should contain empty arrays or None: https://github.com/AMICI-dev/AMICI/blob/master/src/rdata.cpp
> 
> That is also where I looked and what I thought, but as the dimensions of each fields were the same that would not resolve the issue would it?
> 
> Here is a terminal output: `rdata_complete` had the solver set the `RDataReporting` to 0, while `data_in_chunks` had it set to 2.
> 
> ```
> (Pdb) rdata_complete['rdatas'][0]._field_dimensions
> {'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
> (Pdb) rdata_in_chunks['rdatas'][0]._field_dimensions
> {'ts': [10], 'x': [10, 4], 'x0': [4], 'x_ss': [4], 'sx': [10, 29, 4], 'sx0': [29, 4], 'sx_ss': [29, 4], 'y': [10, 5], 'sigmay': [10, 5], 'sy': [10, 29, 5], 'ssigmay': [10, 29, 5], 'z': [10, 0], 'rz': [10, 0], 'sigmaz': [10, 0], 'sz': [10, 29, 0], 'srz': [10, 29, 0], 'ssigmaz': [10, 29, 0], 'sllh': [29], 's2llh': [29, 29], 'res': [50], 'sres': [50, 29], 'FIM': [29, 29], 'J': [4, 4], 'w': [10, 10], 'xdot': [4], 'preeq_numlinsteps': [0, 2], 'preeq_numsteps': [1, 3], 'preeq_status': [1, 3], 'posteq_numlinsteps': [0, 2], 'posteq_numsteps': [1, 3], 'posteq_status': [1, 3], 'numsteps': [10], 'numrhsevals': [10], 'numerrtestfails': [10], 'numnonlinsolvconvfails': [10], 'order': [10], 'numstepsB': [10], 'numrhsevalsB': [10], 'numerrtestfailsB': [10], 'numnonlinsolvconvfailsB': [10]}
> (Pdb)
> ```

Ah, those dimensions are statically set and don't necessarily agree with actual field dimensions. I double checked and with amici 0.11.15 and `reporting_mode = amici.RDataReporting.likelihood`, rdatas[0].res is None.